flytekit/core/promise.py

```diff
@@ -343,7 +407,7 @@ def val(self) -> _literals_models.Literal:
         return self._val
 
     @property
-    def ref(self) -> NodeOutput:
+    def ref(self) -> Optional[NodeOutput]:
```

Comment:
I think it's okay to leave this as before... exceptions don't need to be in the type hint.  so before, the logic was basically, if you have a ref, then it's a node output.  If you don't, and you try to call .ref, then you get an AttributeError.